metrics/internal/cache/pv.go

```diff
@@ -0,0 +1,152 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+// Cache metadata for all PV with annotation - pv.kubernetes.io/provisioned-by: rook-ceph.rbd.csi.ceph.com
+
+var _ cache.Store = &PersistentVolumeStore{}
+
+// PersistentVolumeStore implements the k8s.io/client-go/tools/cache.Store
+// interface. It stores persistent volume CSI attributes
+type PersistentVolumeStore struct {
+	Mutex sync.RWMutex
+	// Store is a map of PV UID to PersistentVolumeAttributes
+	Store map[types.UID]PersistentVolumeAttributes
+}
+
+type PersistentVolumeAttributes struct {
+	PersistentVolumeName           string
+	PersistentVolumeClaimName      string
+	PersistentVolumeClaimNamespace string
+	ImageName                      string
+	Pool                           string
+}
+
+func NewPersistentVolumeStore() *PersistentVolumeStore {
+	return &PersistentVolumeStore{
+		Store: map[types.UID]PersistentVolumeAttributes{},
+	}
+}
+
+// Add inserts to the PersistentVolumeStore.
+func (s *PersistentVolumeStore) Add(obj interface{}) error {
+	o, err := meta.Accessor(obj)
+	if err != nil {
+		return err
+	}
+
+	pv, ok := obj.(*corev1.PersistentVolume)
+	if !ok {
+		return fmt.Errorf("unexpected object of type %T", obj)
+	}
+
+	provisioner := pv.Annotations["pv.kubernetes.io/provisioned-by"]
+	if !strings.Contains(provisioner, ".rbd.csi.ceph.com") {
+		klog.Infof("Skipping non Ceph CSI RBD volume %s", pv.Name)
+		return nil
+	}
+
+	s.Mutex.Lock()
+	defer s.Mutex.Unlock()
+
+	s.Store[o.GetUID()] = PersistentVolumeAttributes{
+		PersistentVolumeName:           pv.Spec.CSI.VolumeAttributes["csi.storage.k8s.io/pv/name"],
+		PersistentVolumeClaimName:      pv.Spec.CSI.VolumeAttributes["csi.storage.k8s.io/pvc/name"],
+		PersistentVolumeClaimNamespace: pv.Spec.CSI.VolumeAttributes["csi.storage.k8s.io/pvc/namespace"],
```

Comment:
we will not have these details in VolumeAttributes anymore with new cephcsi changes, read it from the `spec.claimRef``

Reply:
I was wondering why we had duplicate fields. Will update it.